### DISCUSSION TITLE
Official Instrumentation in Python 3 + Python as a scripting language for Bochs debugger.

### DISCUSSION THREAD

**Discussion by therealdreg** [bochs-emu/Bochs#62] [Polls]
Currently, creating instrumentation in Bochs is a somewhat slow process (bugs/stupid errors)...

Having the ability to use a high-level language like Python for prototyping and 'rapid' development would be advantageous, even though it's understood that Python-based instrumentation would be slower than C code.

 Additionally, the Bochs debugger lacks a scripting language (unlike GDB, which has Python). Its command line resembles a basic calculator, implemented in flex + bison. 

I'm not looking to reinvent the wheel or create an own scripting language (dealing with ambiguous grammars sucks). 

Therefore, the Bochs debugger should feature a scripting language that is easily debuggable. Embedding Python is quite straightforward, and those scripts can later be debugged with pdb: https://docs.python.org/3/library/pdb.html. 

At present, the workflow is slow and tedious; for anything advanced or custom, you must code it in C, compile it, etc. 

Another option could be a JavaScript engine (Windbg-style): Quickjs https://bellard.org/quickjs/. 

However, considering the community support, available libraries, and other factors, Python might be more appealing.

Indeed, the underlying principle of this proposal is to ensure that all new features are optional and determined at compile time. This design consideration allows for customization based on individual preferences or requirements, without any negative impact on existing functionality or performance. In essence, any components that are not needed or desired can be omitted, and the system will continue to operate as it has previously, maintaining the same level of performance.

I could prepare a PR with all these changes. However, this requires a substantial series of modifications to the Debugger layer. Currently, the UI is tightly coupled with the functionality, and several changes would be needed. 

Still, I'm willing to take it on if you think it's worthwhile. Please let me know your thoughts, as I don't want to spend weeks or months on this only for it to end up unused. Thank you!

**Comment by stlintel** [comment]
In general, it is good idea. 
In practice I don't know how to approach to it.
Not every guest has python and not everybody interested in it so it must be completely optional and probably also host specific...
But if done right I believe it could be powerful tool.
BTW, I saw few Bochs forks with python integrated to debugger and instrumentation, I hope you are familiar with them.
One of them was even developed by google project zero team.

**Reply by therealdreg** [reply]
It's not a problem; everything will be optional through #ifdefs. This is mainly geared towards researchers / sandbox / automate. Its not a generic multi-platform feature. So the idea is to have it work on Linux, Windows and Mac.

**Comment by stlintel** [comment]
https://github.com/zynamics/bochs-python-instrumentation ?

one not very related but you might be interested: https://github.com/googleprojectzero/bochspwn

**Reply by therealdreg** [reply]
Yes, I'm familiar with those projects. The idea is to do something similar, improve upon it, and officially support it. I believe this is necessary for reversing, malware analysis, exploiting, automation, and so on, but it will take months, and the PR will be huge, And I'm sure we will have very different ideas and many suggestions throughout the process

**Reply by stlintel** [reply]
Cool, not sure I am able to raise it. But if you can try - I will support you

**Comment by impost0r** [comment]
A great idea. I support the idea of Python due to its rich ecosystem and... not being JavaScript. That being said, JS can have its advantages too. I think either way you're going in the right direction, but some ways are better than others (Python).

**Comment by micronn** [comment]
Great idea. I also find it hard to decide between Python or QuickJS. But, after thinking about it for a while, and given that this would be an optional feature, I prefer Python. Sure it would require more than a simple QuickJS, but this comes with functionality that we'll probably want to use anyways (and maybe cleaner syntax too).

**Reply by impost0r** [reply]
Definitely. Hard agree on Python preference. I'm biased as I'm (kind of) a Python developer, but it'd be nice to see anyways. And thinking of the extensibility... it'd be awesome.

**Comment by eqn-grp** [comment]
no kill yourself, use other scripting language like lisp, lua.

**Reply by therealdreg** [reply]
Hey, could you elaborate a bit more on your response? The Python community is vast, and there are a multitude of resources geared towards exploiting, reversing, and other similar areas.

**Comment by monisc-research** [comment]
Strong disagree on the python suggestion. The dependency and versioning nightmare has already been bad enough with IDA - I do not wish to see the same happening on Bochs as well. This wouldn't be the first time JS is used like this - look at frida etc.

However - I do realize that others are more comfortable with using Python due to the ability to reuse dependencies and snippets, so I think a reasonable compromise is having a standard API that is exposed to both JS and Python. This also opens up the possibility of exposing the API to WASM too - allowing the use of Rust, C, etc as a scripting language.

I would be more than glad to help out with the QuickJS integration into Bochs as that was on my todo list anyways.

**Reply by therealdreg** [reply]
Alright, once the refactoring and redesign of the necessary components for this milestone are completed, I'll handle it in Python. Will you take care of the QuickJS part?

**Reply by monisc-research** [reply]
Yep, let me know when that is done and I'll get started.
